vllm/engine/async_llm_engine.py

```diff
@@ -466,15 +466,17 @@ async def add_request_async(
         )
         processed_inputs = self.input_processor(preprocessed_inputs)
 
-        if isinstance(params, SamplingParams):
+        if isinstance(params, SamplingParams) and \
+            params.guided_decoding is not None:
             # Guided decoding has an async implementation for building logits
             # processors in a separate threadpool.
             # We want to invoke that here instead of using the blocking
             # implementation in the LLMEngine
-            params = await self._build_guided_decoding_logits_processor_async(
-                params,
-                lora_request,
-            )
+            params = await build_guided_decoding_logits_processor_async(
+                sampling_params=params,
+                tokenizer=self.get_tokenizer(lora_request),
```

Comment:
```suggestion
                tokenizer=await self.get_tokenizer(lora_request),
```

But actually I was thinking could pass the function itself here (along with lora_request I guess) so that it can be awaited inside the model if needed...

Reply:
We talked- the reason this is the way it is is that the inner `_AsyncLLMEngine` class doesn't have an async get_tokenizer method, so this resolves to the `LLMEngine` def.

This might be a small potential performance issue, but this only affects users of the AsyncLLMEngine, and current usage won't be affected since these are new parameters in `SamplingParams`